datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -767,6 +777,92 @@ private Future<Boolean> dispatchAssignmentChangeIfNeeded(String connectorType, L
+    Set<DatastreamTask> removedTasks = new HashSet<>(oldAssignment);
+    newAssignment.forEach(removedTasks::remove);
+    return new ArrayList<>(removedTasks);
+  }
+
+  private List<DatastreamTask> getAddedTasks(List<DatastreamTask> newAssignment, List<DatastreamTask> oldAssignment) {
+    Set<DatastreamTask> addedTasks = new HashSet<>(newAssignment);
+    oldAssignment.forEach(addedTasks::remove);
+    return new ArrayList<>(addedTasks);
+  }
+
+  @VisibleForTesting
+  void maybeClaimAssignmentTokensForStoppingStreams(List<DatastreamTask> newAssignment,
+      List<DatastreamTask> oldAssignment) {
+    // TODO Add metrics for number of streams that are inferred as stopping
+    Map<String, List<DatastreamTask>> newAssignmentPerConnector = new HashMap<>();
+    for (DatastreamTask task : newAssignment) {
+      String connectorType = task.getConnectorType();
+      newAssignmentPerConnector.computeIfAbsent(connectorType, k -> new ArrayList<>()).add(task);
+    }
+
+    Map<String, List<DatastreamTask>> oldAssignmentPerConnector = new HashMap<>();
+    for (DatastreamTask task : oldAssignment) {
+      String connectorType = task.getConnectorType();
+      oldAssignmentPerConnector.computeIfAbsent(connectorType, k -> new ArrayList<>()).add(task);
+    }
+
+    Set<String> allConnectors = new HashSet<>();
+    allConnectors.addAll(newAssignmentPerConnector.keySet());
+    allConnectors.addAll(oldAssignmentPerConnector.keySet());
+
+    // The follower nodes don't keep an up-to-date view of the datastreams Zk node and making them listen to changes
+    // to datastreams node risks overwhelming the ZK server with reads. Instead, the follower is inferring the stopping
+    // streams from the task assignment for each connector type. Even if that inference results in a falsely identified
+    // stopping stream, the attempt to claim the token will be a no-op.
+    allConnectors.parallelStream().forEach(connector -> {
+      List<DatastreamTask> oldTasks = oldAssignmentPerConnector.getOrDefault(connector, Collections.emptyList());
+      List<DatastreamTask> newTasks = newAssignmentPerConnector.getOrDefault(connector, Collections.emptyList());
+      List<DatastreamTask> removedTasks = getRemovedTasks(newTasks, oldTasks);
+
+      List<Datastream> stoppingStreams = inferStoppingDatastreamsFromAssignment(newTasks, removedTasks);
+      Set<String> stoppingStreamNames = stoppingStreams.stream().map(Datastream::getName).collect(Collectors.toSet());
+
+      if (!stoppingStreamNames.isEmpty()) {
+        _log.info("Trying to claim assignment tokens for connector {}, streams: {}", connector, stoppingStreamNames);
+
+        Set<String> stoppingDatastreamTasks = removedTasks.stream().
+            filter(t -> stoppingStreamNames.contains(t.getTaskPrefix())).
+            map(DatastreamTask::getId).collect(Collectors.toSet());
+
+        if (PollUtils.poll(() -> connectorTasksHaveStopped(connector, stoppingDatastreamTasks),
```

Comment:
Should we instead put this in a loop for each datastream? Not a strong requirement, but this will prevent having dependencies on other stopping streams for a particular stream to stop.

Reply:
It feels like a micro-optimization. The poll is already parallel for connectors, and also, this whole method is scheduled to run asynchronously in `handleAssignmentChange`.

Reply:
In some scenarios, a stuck pending-to-stop task for a particular stream could block a successful stopping datastream for the same connector, hence I suggested that.

Reply:
Stop requests usually affect one stream, there's no support for bulk stops/restarts, so in production scenarios multiple stopping streams (or rather, streams that are inferred as stopping) are virtually unexpected. I can't think of a scenario that would trigger a bulk stop/restart on the server. 
It's a tradeoff of complicating the code and the value we get out of it. It's not hard for me to add it, I'll let @vmaheshw be the tie breaker here.

Reply:
Parallel restarts, or stopping all the datastreams, usually happens during outage recovery. In that scenario, if the stopping group of datastreams is stuck because of one, do we have an alternate way to recover (maybe force-stopping the datastream)?

I agree with both of you here. Let's add this as comments to start with in case this becomes a pain in the future. We can implement it then.